controllers/profilecatalogsource_controller.go

```diff
@@ -74,32 +90,91 @@ func (r *ProfileCatalogSourceReconciler) Reconcile(ctx context.Context, req ctrl
 		return ctrl.Result{}, err
 	}
 
-	logger.Info("updating catalog entries", "profiles", pCatalog.Spec.Profiles)
-	r.Profiles.Update(pCatalog.Name, pCatalog.Spec.Profiles...)
+	//can configre spec.Profiles or spec.Repositories, not both.
+	if len(pCatalog.Spec.Profiles) > 0 {
+		logger.Info("updating catalog entries", "profiles", pCatalog.Spec.Profiles)
+		r.Profiles.AddOrReplace(pCatalog.Name, pCatalog.Spec.Profiles...)
+		return ctrl.Result{}, nil
+	}
+
+	gitRepoManager := gitrepository.NewManager(ctx, pCatalog.Namespace, r.Client, r.timeout, r.interval)
+	scanner := r.newScanner(gitRepoManager, &git.Client{}, http.DefaultClient, logger)
+	catalogExists := r.Profiles.CatalogExists(pCatalog.Name)
 
-	timeout := time.Minute * 2
-	interval := time.Second * 5
-	gitRepoManager := gitrepository.NewManager(ctx, pCatalog.Namespace, r.Client, timeout, interval)
-	scanner := scanner.New(gitRepoManager, &git.Client{}, http.DefaultClient, logger)
 	for _, repo := range pCatalog.Spec.Repos {
 		logger.Info("scan repo for profiles", "repo", repo)
 		var secret *corev1.Secret
 		if repo.SecretRef != nil {
 			secret = &corev1.Secret{}
-			objectKey := client.ObjectKey{Name: repo.SecretRef.Name, Namespace: req.Namespace}
+			objectKey := client.ObjectKey{Name: repo.SecretRef.Name, Namespace: pCatalog.Namespace}
 			if err := r.Client.Get(ctx, objectKey, secret); err != nil {
 				return ctrl.Result{}, fmt.Errorf("failed to find secret for repo %v: %w", repo, err)
 			}
 		}
 
-		profiles, err := scanner.ScanRepository(repo, secret)
+		var alreadyScannedTags []string
+		if catalogExists {
+			for _, scannedRepo := range pCatalog.Status.ScannedRepositories {
+				if scannedRepo.URL == repo.URL {
+					alreadyScannedTags = scannedRepo.Tags
+				}
+			}
+		}
+
+		profiles, newTags, err := scanner.ScanRepository(repo, secret, alreadyScannedTags)
 		if err != nil {
 			return ctrl.Result{}, err
 		}
-		r.Profiles.Update(pCatalog.Name, profiles...)
+
+		if catalogExists {
+			createOrAppendScannedRepositoryStatus(&pCatalog, repo, newTags)
```

Comment:
These two functions are doing the same thing. I think it would be a lot nicer to have a single function with a parameter called `update`. Which if true, then just append. Otherwise, let it do its thing. So like:

```go
func updateScannedRepositoryStatus(pCatalog *profilesv1.ProfileCatalogSource, repo profilesv1.Repository, newTags []string, replace bool) {
	for i, scannedRepo := range pCatalog.Status.ScannedRepositories {
		if scannedRepo.URL == repo.URL {
                         if replace {
			      pCatalog.Status.ScannedRepositories[i].Tags = newTags
                         } else {
                               pCatalog.Status.ScannedRepositories[i].Tags = append(scannedRepo.Tags, newTags...)
                         }
			return
		}
	}
	pCatalog.Status.ScannedRepositories = append(pCatalog.Status.ScannedRepositories, profilesv1.ScannedRepository{
		URL:  repo.URL,
		Tags: newTags,
	})
}
```

So you can remove the other function. WDYT?

Reply:
done!